modules/test/portlet-container-test/test/integration/com/liferay/portlet/container/test/BaseTestCase.java

```diff
@@ -0,0 +1,243 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portlet.container.test;
+
+import com.liferay.portal.kernel.test.rule.DeleteAfterTestRun;
+import com.liferay.portal.kernel.test.util.GroupTestUtil;
+import com.liferay.portal.kernel.test.util.TestPropsValues;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.portal.model.Company;
+import com.liferay.portal.model.Group;
+import com.liferay.portal.model.Layout;
+import com.liferay.portal.model.Portlet;
+import com.liferay.portal.security.permission.PermissionChecker;
+import com.liferay.portal.security.permission.PermissionCheckerFactoryUtil;
+import com.liferay.portal.security.permission.PermissionThreadLocal;
+import com.liferay.portal.service.CompanyLocalServiceUtil;
+import com.liferay.portal.theme.ThemeDisplay;
+import com.liferay.portal.theme.ThemeDisplayFactory;
+import com.liferay.portal.util.test.LayoutTestUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import java.util.Arrays;
+import java.util.Dictionary;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.util.tracker.ServiceTracker;
+
+import org.springframework.mock.web.MockHttpServletRequest;
+
+/**
+ * @author Raymond Augé
+ */
+public class BaseTestCase {
+
+	@Before
+	public void setUp() throws Exception {
+		group = GroupTestUtil.addGroup();
+
+		layout = LayoutTestUtil.addLayout(group);
+
+		PermissionChecker permissionChecker =
+			PermissionCheckerFactoryUtil.create(TestPropsValues.getUser());
+
+		PermissionThreadLocal.setPermissionChecker(permissionChecker);
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		PermissionThreadLocal.setPermissionChecker(null);
+
+		if (serviceTracker != null) {
+			serviceTracker.close();
+		}
+
+		for (ServiceRegistration<?> serviceRegistration :
+				serviceRegistrations) {
+
+			serviceRegistration.unregister();
+		}
+
+		serviceRegistrations.clear();
+	}
+
+	protected String drain(InputStream stream) throws IOException {
+		if (stream == null) {
+			return "";
+		}
+
+		byte[] bytes = new byte[100];
+		StringBuffer buffer = new StringBuffer(500);
+		int length = 0;
+
+		while ((length = stream.read(bytes)) != -1) {
+			String chunk = new String(bytes, 0, length);
```

Comment:
It's copied from my own code in the OSGi Http whiteboard CT where it's
working perfectly well.

On Tue, Aug 11, 2015 at 7:36 PM, Shuyang Zhou <notifications@github.com>
wrote:

> In
> modules/test/portlet-container-test/test/integration/com/liferay/portlet/container/test/BaseTestCase.java
> <https://github.com/brianchandotcom/liferay-portal/pull/29207#discussion_r36812724>
> :
>
> > +		}
> > +
> > +		serviceRegistrations.clear();
> > +	}
> > +
> > +	protected String drain(InputStream stream) throws IOException {
> > +		if (stream == null) {
> > +			return "";
> > +		}
> > +
> > +		byte[] bytes = new byte[100];
> > +		StringBuffer buffer = new StringBuffer(500);
> > +		int length = 0;
> > +
> > +		while ((length = stream.read(bytes)) != -1) {
> > +			String chunk = new String(bytes, 0, length);
>
> Technically this is wrong, as the byte[] from an InputStream, we don't
> really know it is decodable as String (although you can manually ensure
> that by the test logic), so the new String(bytes, 0, length) could fail on
> decoding.
>
> But that's really not the point, this code really look like a copy paste,
> correct me if it is not true. And it looks pretty out of date. We are
> building some new after all, should not carry on unnecessary baggages.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/brianchandotcom/liferay-portal/pull/29207/files#r36812724>
> .
>



-- 
*Raymond Augé* <http://www.liferay.com/web/raymond.auge/profile>
 (@rotty3000)
Senior Software Architect *Liferay, Inc.* <http://www.liferay.com>
 (@Liferay)
Board Member & EEG Co-Chair, OSGi Alliance <http://osgi.org> (@OSGiAlliance)